### PR TITLE
rewrite vuHandle as a state machine and fix some more races

### PR DESCRIPTION
**Pull request by MStoykov**

### CODE REVIEW

File: lib/executor/vu_handle.go [hunk 1]
```diff
@@ -84,20 +93,24 @@ func newStoppedVUHandle(
 			// we are done just ruturn the VU
 			vh.initVU = nil
 			vh.activeVU = nil
-			atomic.StoreInt32(&vh.change, 1)
+			atomic.StoreInt32(&vh.state, stopped) // change this to soemthign that exits ?
```
Comment (na--) [inline lib/executor/vu_handle.go:96]:
```suggestion
			atomic.StoreInt32(&vh.state, stopped) // change this to something that exits ?
```

File: lib/executor/vu_handle.go [hunk 2]
```diff
@@ -77,52 +87,51 @@ func newStoppedVUHandle(
...
-	default:
+	if vh.state == toGracefulStop { // No point returning the VU just continue as it was
 		close(vh.canStartIter)
+		vh.changeState(running)
+		vh.mutex.Unlock()
+		return
 	}
+
+	if vh.state == starting || vh.state == running {
+		vh.mutex.Unlock()
```
Comment (imiric) [inline lib/executor/vu_handle.go:122]: Could we `defer vh.mutex.Unlock()` at the start to avoid repetition?

File: lib/executor/vu_handle.go [hunk 3]
```diff
@@ -77,52 +87,51 @@ func newStoppedVUHandle(
...
 }
 
+// start is one of the more complicated ones as it can race with toGracefulStop and has special
+// cases
+// state changes
+// toGracefulStop => running, we have raced with toGracefulStop if it was finished the state
+// would've been stoppped
+// starting, running => nothing, we have done what needs to be done
+// stopped, toHardStop => reactivate everything, it is either stopped or hardStop has uninitialized
+// some stuff and we need to reinitialize them
```
Comment (imiric) [inline lib/executor/vu_handle.go:111]: Would you mind creating a [state-transition table](https://en.wikipedia.org/wiki/State-transition_table) (or a state diagram in something like graphviz/DOT if you're inspired :smile:) for this? That way one can reference it for all methods, and it's a bit easier to read.

File: lib/executor/vu_handle.go (hunk 2 above)
Comment (MStoykov) [reply lib/executor/vu_handle.go:122]: yeah ... I start wondering the same as well .. it won't be possible in `runLoopsIfPossible`. but I would argue the slowdown from defer (which now is way smaller) is not such a problem for this methods ...

File: lib/executor/vu_handle.go (hunk 3 above)
Comment (MStoykov) [reply lib/executor/vu_handle.go:111]: I had a drawing (on paper with pen), which got outdated ... way too long ago, and I intented on having it as a comment in ascii art, but a state-transition table seems better

Comment (MStoykov) [reply lib/executor/vu_handle.go]: fixed ?

File: lib/executor/vu_handle.go (hunk 2 above)
Comment (MStoykov) [reply lib/executor/vu_handle.go]: fixed in 337fc21

File: lib/executor/vu_handle.go (hunk 3 above)
Comment (imiric) [reply lib/executor/vu_handle.go]: Yeah, thanks, that's much easier to understand now. I think the transitions make sense, but will have to dig a bit deeper.

**Pull request merged**